airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/components/debezium/StateFilesAccessor.kt

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+package io.airbyte.cdk.components.debezium
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.google.common.collect.ImmutableList
+import io.airbyte.commons.jackson.MoreMappers
+import io.debezium.config.Configuration
+import io.debezium.relational.history.AbstractFileBasedSchemaHistory
+import io.debezium.relational.history.HistoryRecord
+import io.debezium.relational.history.HistoryRecordComparator
+import io.debezium.relational.history.SchemaHistoryListener
+import io.debezium.storage.file.history.FileSchemaHistory
+import java.io.IOException
+import java.io.UncheckedIOException
+import java.lang.reflect.InvocationTargetException
+import java.lang.reflect.Method
+import java.nio.ByteBuffer
+import java.nio.file.Files
+import java.nio.file.Path
+import java.util.function.Consumer
+import kotlin.Exception
+import kotlin.RuntimeException
+import kotlin.String
+import org.apache.commons.io.FileUtils
+import org.apache.kafka.common.utils.ByteBufferInputStream
+import org.apache.kafka.connect.json.JsonConverter
+import org.apache.kafka.connect.json.JsonConverterConfig
+import org.apache.kafka.connect.runtime.WorkerConfig
+import org.apache.kafka.connect.runtime.standalone.StandaloneConfig
+import org.apache.kafka.connect.storage.Converter
+import org.apache.kafka.connect.storage.FileOffsetBackingStore
+
+/**
+ * [StateFilesAccessor] reads and writes from Debezium offset and schema history files. This should
+ * NOT be done while the Debezium Engine is running! None of this is thread-safe. Instead, this
+ * object should be used to:
+ * 1. create and populate the files prior to running the Debezium Engine,
+ * 2. read the changed files after running the Debezium Engine.
+ *
+ * The files are deleted on [close].
+ */
+class StateFilesAccessor : AutoCloseable {
+
+    val offsetFilePath: Path
+    val schemaFilePath: Path
+
+    internal val workingDir: Path
+    private val fileOffsetBackingStore: FileOffsetBackingStore
+    private val fileSchemaHistory: FileSchemaHistory
+
+    override fun close() {
+        fileOffsetBackingStore.stop()
+        fileSchemaHistory.stop()
+        try {
+            FileUtils.deleteDirectory(workingDir.toFile())
+        } catch (e: IOException) {
+            throw UncheckedIOException(e)
+        }
+    }
+
+    fun readUpdatedOffset(offset: DebeziumComponent.State.Offset): DebeziumComponent.State.Offset {
+        val bytesKeys = offset.debeziumOffset.keys.map { key -> toByteBuffer(key) }
+        fileOffsetBackingStore.start()
+        val future = fileOffsetBackingStore.get(bytesKeys)
+        fileOffsetBackingStore.stop()
+        val debeziumOffset =
+            future.resultNow().map { (key, value) -> Pair(toJson(key), toJson(value)) }.toMap()
+        return DebeziumComponent.State.Offset(debeziumOffset)
+    }
+
+    fun writeOffset(offset: DebeziumComponent.State.Offset) {
```

Comment:
I didn't find where we define this state - would that just be a big json/map covering for all cases among dbs?

Reply:
This state is either defined using an airbyte state message or synthetized from scratch. It's not super straightforward to find that logic, it's in the top PR in the stack: https://github.com/airbytehq/airbyte/pull/35889/files#diff-e8298352c23705a30cdea980917f1a8cccef801e79c08da2445802927d4ad04aR78-R104
In fact, this is just a rewrite of the existing `constructInitialDebeziumState` method in `PostgresDebeziumStateUtil`.

It consists of a map where the key identifies the db and the value has the actual offset info. In our usage of debezium, we only ever have one key.